tests/phpunit/Civi/Token/TokenCompatSubscriberTest.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Civi\Token;
+
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+class TokenCompatSubscriberTest extends \CiviUnitTestCase {
+
+  /**
+   * @var \Symfony\Component\EventDispatcher\EventDispatcher
+   */
+  protected $dispatcher;
+
+  protected function setUp() {
+    $this->useTransaction(TRUE);
+    parent::setUp();
+    $this->dispatcher = new EventDispatcher();
+    $tcs = new TokenCompatSubscriber();
+    $this->dispatcher->addListener(Events::TOKEN_EVALUATE, [
+      $tcs,
+      'onEvaluate',
+    ]);
+    $this->dispatcher->addListener(Events::TOKEN_RENDER, [$tcs, 'onRender']);
+    $this->hookClass = \CRM_Utils_Hook::singleton();
+  }
+
+  /**
+   *
+   */
+  public function testGreetingToken() {
+    $cid = $this->individualCreate([
+      'last_name' => 'Tester',
+      'first_name' => 'Cata',
+    ]);
+
+    $p = new TokenProcessor($this->dispatcher, [
+      'controller' => __CLASS__,
+    ]);
+    $p->addMessage('greeting_html', '<p>{contact.email_greeting}</p>. {custom.foobar} Bye!', 'text/html');
+    $p->addMessage('greeting_text', '{contact.email_greeting}. {custom.foobar} Bye!', 'text/plain');
+    $p->addRow()
+      ->context(['contactId' => $cid]);
+
+    $expectHtml = [
+      0 => '<p>Dear Cata</p>.  Bye!',
+    ];
+
+    $expectText = [
+      0 => 'Dear Cata.  Bye!',
+    ];
+
+    foreach ($p->evaluate()->getRows() as $key => $row) {
+      $this->assertEquals($expectHtml[$key], $row->render('greeting_html'));
+      $this->assertEquals($expectText[$key], $row->render('greeting_text'));
```

Comment:
Small/non-blocking suggestion - when doing assertions over an array, I've seen a few situations where an empty array yielded a false-pass. Adding some kind of assertion on the count can hedge that, e.g.

```php
$p->evaluate();
$this->assertEquals(1, count($p->getRows()));
foreach ($p->getRows() as $key => $row) {
       $this->assertEquals($expectHtml[$key], $row->render('greeting_html'));
```

Reply:
@totten thanks for pointing out. I will keep this one as it is now.